### PR TITLE
Fix up HTTP message framing edge cases.

### LINKED ISSUE #254: HTTP/1 parser should handle semantic EOF.
Right now the HTTP/1 decoder does not correctly handle EOF. EOF can potentially be semantic in HTTP/1.1, and so we should correctly pass it to HTTPParser when we see it. Right now we don't. I bumped into this while working on an unrelated patch so I don't plan to handle it immediately, but I wanted to file this issue so I didn't forget.

### PR DESCRIPTION
**Pull request by Lukasa**
Motivation:

HTTP message framing has a number of edge cases that NIO currently does
not tolerate. We should decide what our position is on each of these edge
cases and handle it appropriately.

Modifications:

Provide an extensive test suite that codifies our expected handling of
these edge cases. Fix divergences from this behaviour.

Result:

Better tolerance for the weird corners of HTTP.

Resolves #254.

### CODE REVIEW

File: Sources/NIOHTTP1/HTTPDecoder.swift [hunk 1]
```diff
@@ -263,14 +268,34 @@ public class HTTPDecoder<HTTPMessageT>: ByteToMessageDecoder, AnyHTTPDecoder {
...
+                // codes and override http_parser's handling as well.
                 let method = handler.popRequestMethod()
-                return method == .HEAD ? 1 : 0
+                if method == .HEAD || method == .CONNECT {
+                    return 1
+                }
+
+                if ((head.status.code / 100 == 1) ||  // 1XX codes
+                    (head.status.code == 204) ||
+                    (head.status.code == 304)) {
```
Comment (normanmaurer) [inline Sources/NIOHTTP1/HTTPDecoder.swift]: nit: You don't need the `(...)` around rich of these

File: Sources/NIOHTTP1/HTTPDecoder.swift [hunk 2]
```diff
@@ -437,6 +469,40 @@ public class HTTPDecoder<HTTPMessageT>: ByteToMessageDecoder, AnyHTTPDecoder {
...
+            return
+        }
+
+        // If we got here, no error was hit, but we may have a pending callout. Let's fire it if we did.
+        if self.pendingInOut.count > 0 {
+            self.channelReadComplete(ctx: ctx)
+        }
+
+        // We don't need the cumulation buffer, if we're holding it.
+        self.cumulationBuffer = nil
```
Comment (normanmaurer) [inline Sources/NIOHTTP1/HTTPDecoder.swift]: @Lukasa should we set this to nil before we call `channelReadComplete` to reduce the reference count and so minimize the chances of CoW ?

File: Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+    var seenHead = false
+
+    func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
+        switch self.unwrapInboundIn(data) {
+        case .head:
+            self.seenHead = true
+        case .body:
+            self.seenBody = true
+        case .end:
+            self.seenEnd = true
```
Comment (normanmaurer) [inline Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift]: should we also assert that we not seen the same state before ?

File: Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift [hunk 4]
```diff
@@ -0,0 +1,519 @@
...
+            channel.pipeline.fireUserInboundEventTriggered(ChannelEvent.inputClosed)
+        } else {
+            channel.pipeline.fireChannelInactive()
+        }
+        XCTAssertNotNil(handler.response)
+        XCTAssertEqual(handler.body, Array("some body data".utf8))
+        XCTAssertTrue(handler.receivedEnd)
+        XCTAssertTrue(handler.eof)
+
+        XCTAssertNoThrow(try channel.finish())
```
Comment (normanmaurer) [inline Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift]: I think we should actually assert the return value of channel.finish()

File: Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift [hunk 5]
```diff
@@ -0,0 +1,519 @@
...
+        response.write(staticString: "\r\n")
+
+        XCTAssertNoThrow(try channel.writeInbound(IOData.byteBuffer(response)))
+
+        // We should have a response, no body, and immediately see EOF.
+        XCTAssert(handler.seenHead)
+        XCTAssertFalse(handler.seenBody)
+        XCTAssert(handler.seenEnd)
+
+        XCTAssertNoThrow(try channel.finish())
```
Comment (normanmaurer) [inline Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift]: I think we should actually assert the return value of channel.finish()

File: Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift [hunk 6]
```diff
@@ -0,0 +1,519 @@
...
+
+        // Send a GET with the appropriate Transfer Encoding header.
+        XCTAssertNoThrow(try channel.writeInbound(ByteBuffer(string: "POST / HTTP/1.1\r\nTransfer-Encoding: \(transferEncodingHeader)\r\n\r\n")))
+
+        // We should have a request, no body, and immediately see end of request.
+        XCTAssert(handler.seenHead)
+        XCTAssertFalse(handler.seenBody)
+        XCTAssert(handler.seenEnd)
+
+        XCTAssertNoThrow(try channel.finish())
```
Comment (normanmaurer) [inline Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift]: I think we should actually assert the return value of channel.finish()

File: Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift [hunk 7]
```diff
@@ -0,0 +1,519 @@
...
+        if case .halfClosure = eofMechanism {
+            channel.pipeline.fireUserInboundEventTriggered(ChannelEvent.inputClosed)
+        } else {
+            channel.pipeline.fireChannelInactive()
+        }
+        XCTAssert(handler.seenHead)
+        XCTAssert(handler.seenBody)
+        XCTAssert(handler.seenEnd)
+
+        XCTAssertNoThrow(try channel.finish())
```
Comment (normanmaurer) [inline Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift]: I think we should actually assert the return value of channel.finish()

File: Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift [hunk 8]
```diff
@@ -0,0 +1,519 @@
...
+            try channel.writeInbound(ByteBuffer(string: "POST / HTTP/1.1\r\nTransfer-Encoding: chunked\r\nContent-Length: 4\r\n\r\n"))
+            XCTFail("Did not throw")
+        } catch HTTPParserError.unexpectedContentLength {
+            // ok
+        } catch {
+            XCTFail("Unexpected error: \(error)")
+        }
+
+        // Must spin the loop.
+        (channel.eventLoop as! EmbeddedEventLoop).run()
```
Comment (normanmaurer) [inline Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift]: Also call `channel.finish()` and assert return value.

Comment (normanmaurer) [inline Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift]: Should we also add `XCTAssertFalse(channel.isActive` ?

File: Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift [hunk 9]
```diff
@@ -0,0 +1,519 @@
...
+            try channel.writeInbound(ByteBuffer(string: "HTTP/1.1 200 OK\r\nTransfer-Encoding: chunked\r\nContent-Length: 4\r\n\r\n"))
+            XCTFail("Did not throw")
+        } catch HTTPParserError.unexpectedContentLength {
+            // ok
+        } catch {
+            XCTFail("Unexpected error: \(error)")
+        }
+
+        // Must spin the loop.
+        (channel.eventLoop as! EmbeddedEventLoop).run()
```
Comment (normanmaurer) [inline Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift]: I think we should actually assert the return value of channel.finish()

Comment (normanmaurer) [inline Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift]: Assert that the channel is not active anymore

File: Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift [hunk 10]
```diff
@@ -0,0 +1,519 @@
...
+        } catch HTTPParserError.invalidContentLength {
+            // ok
+        } catch HTTPParserError.unexpectedContentLength {
+            // also ok
+        } catch {
+            XCTFail("Unexpected error: \(error)")
+        }
+
+        // Must spin the loop.
+        (channel.eventLoop as! EmbeddedEventLoop).run()
```
Comment (normanmaurer) [inline Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift]: assert the channel is not native anymore.

Comment (normanmaurer) [reply Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift]: I think we should actually assert the return value of channel.finish()

File: Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift [hunk 11]
```diff
@@ -0,0 +1,519 @@
...
+            try channel.writeInbound(ByteBuffer(string: "POST / HTTP/1.1\r\nContent-Length: 4, 4\r\n\r\n"))
+            XCTFail("Did not throw")
+        } catch HTTPParserError.invalidContentLength {
+            // ok
+        } catch {
+            XCTFail("Unexpected error \(error)")
+        }
+
+        // Must spin the loop.
+        (channel.eventLoop as! EmbeddedEventLoop).run()
```
Comment (normanmaurer) [inline Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift]: assert the channel is not active anymore

Comment (normanmaurer) [reply Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift]: I think we should actually assert the return value of channel.finish()

File: Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift [hunk 12]
```diff
@@ -0,0 +1,519 @@
...
+        XCTAssertNoThrow(try channel.pipeline.add(handler: handler).wait())
+
+        // Send a POST without a length field of any kind. This should be a zero-length request,
+        // so .end should come immediately.
+        XCTAssertNoThrow(try channel.writeInbound(ByteBuffer(string: "POST / HTTP/1.1\r\nHost: example.org\r\n\r\n")))
+        XCTAssert(handler.seenHead)
+        XCTAssertFalse(handler.seenBody)
+        XCTAssert(handler.seenEnd)
+
+        XCTAssertNoThrow(try channel.finish())
```
Comment (normanmaurer) [inline Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift]: assert the return value

File: Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift [hunk 13]
```diff
@@ -0,0 +1,519 @@
...
+            try channel.writeInbound(ByteBuffer(string: "POST / HTTP/1.1\r\nContent-Length: 4\r\nContent-Length: 4\r\n\r\n"))
+            XCTFail("Did not throw")
+        } catch HTTPParserError.unexpectedContentLength {
+            // ok
+        } catch {
+            XCTFail("Unexpected error \(error)")
+        }
+
+        // Must spin the loop.
+        (channel.eventLoop as! EmbeddedEventLoop).run()
```
Comment (normanmaurer) [inline Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift]: assert that the channel is not active anymore and also call `channel.finish()` and assert return value.

File: Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift (hunk 8 above)
Comment (Lukasa) [reply Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift]: Nope, we can't do that here because `channel.finish()` will throw if the channel is already closed.

File: Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift (hunk 9 above)
Comment (Lukasa) [reply Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift]: Nope, we can't do that here because `channel.finish()` will throw if the channel is already closed.

File: Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift (hunk 10 above)
Comment (Lukasa) [reply Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift]: Nope, we can't do that here because `channel.finish()` will throw if the channel is already closed.

File: Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift (hunk 11 above)
Comment (Lukasa) [reply Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift]: Nope, we can't do that here because `channel.finish()` will throw if the channel is already closed.

File: Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift (hunk 13 above)
Comment (Lukasa) [reply Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift]: Nope, we can't do that here because `channel.finish()` will throw if the channel is already closed.

File: Sources/NIOHTTP1/HTTPDecoder.swift (hunk 2 above)
Comment (Lukasa) [reply Sources/NIOHTTP1/HTTPDecoder.swift]: Sure, in fact we can set it to `nil` very early in this function without danger.

**Comment by Lukasa** [comment]
Ok @normanmaurer, please re-review.

File: Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift [hunk 14]
```diff
@@ -0,0 +1,530 @@
...
+        case .end:
+            XCTAssertFalse(self.seenEnd)
+            self.seenEnd = true
+        }
+    }
+}
+
+/// Tests for the HTTP decoder's handling of message body framing.
+///
+/// Mostly tests assertions in RFC 7230 § 3.3.3.
```
Comment (normanmaurer) [inline Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift]: @Lukasa nit: consider linking the rfc directly via markdown. (feel free to ignore tho)

File: Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift [hunk 15]
```diff
@@ -0,0 +1,530 @@
...
+        XCTAssertNoThrow(try channel.pipeline.add(handler: HTTPResponseDecoder()).wait())
+
+        let handler = MessageEndHandler<HTTPResponseHead, ByteBuffer>()
+        XCTAssertNoThrow(try channel.pipeline.add(handler: handler).wait())
+
+        // Prime the decoder with a request and consume it.
+        XCTAssertTrue(try channel.writeOutbound(HTTPClientRequestPart.head(HTTPRequestHead(version: .init(major: 1, minor: 1),
+                                                                                           method: requestMethod,
+                                                                                           uri: "/"))))
+        _ = channel.readOutbound()
```
Comment (normanmaurer) [inline Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift]: should we ensure this not returns nil ?

File: Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift [hunk 16]
```diff
@@ -0,0 +1,530 @@
...
+        XCTAssertNoThrow(try channel.pipeline.add(handler: HTTPResponseDecoder()).wait())
+
+        let handler = MessageEndHandler<HTTPResponseHead, ByteBuffer>()
+        XCTAssertNoThrow(try channel.pipeline.add(handler: handler).wait())
+
+        // Prime the decoder with a request and consume it.
+        XCTAssertTrue(try channel.writeOutbound(HTTPClientRequestPart.head(HTTPRequestHead(version: .init(major: 1, minor: 1),
+                                                                                           method: .GET,
+                                                                                           uri: "/"))))
+        _ = channel.readOutbound()
```
Comment (normanmaurer) [inline Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift]: should we ensure this not return nil ?

**Comment by Lukasa** [comment]
Ok, addressed.

**Comment by Lukasa** [comment]
@swift-nio-bot test this please

**Comment by tomerd** [comment]
@swift-nio-bot test this please

File: Sources/NIOHTTP1/HTTPDecoder.swift [hunk 17]
```diff
@@ -437,6 +464,50 @@ public class HTTPDecoder<HTTPMessageT>: ByteToMessageDecoder, AnyHTTPDecoder {
...
+        self.state.seenEOF = true
+        _ = c_nio_http_parser_execute(&self.parser, &self.settings, nil, 0)
+
+        // We don't need the cumulation buffer, if we're holding it.
+        self.cumulationBuffer = nil
+
+        // No check to state.currentError becuase, if we hit it before, we already threw that
+        // error. This never calls any of the callbacks that set that field anyway. Instead we
+        // just check if the errno is set and throw.
+        let errno = parser.http_errno
```
Comment (weissi) [inline Sources/NIOHTTP1/HTTPDecoder.swift]: I'm not a huge fan of naming variables `errno` because if you rename them and forget the usage sites, it'll still work. Mind making it `httpError` or `httpErrno` or something?

File: Sources/NIOHTTP1/HTTPDecoder.swift [hunk 18]
```diff
@@ -395,7 +416,8 @@ public class HTTPDecoder<HTTPMessageT>: ByteToMessageDecoder, AnyHTTPDecoder {
 
             let errno = parser.http_errno
             if errno != 0 {
-                throw HTTPParserError.httpError(fromCHTTPParserErrno: http_errno(rawValue: errno))!
+                self.state.currentError = HTTPParserError.httpError(fromCHTTPParserErrno: http_errno(rawValue: errno))!
```
Comment (weissi) [inline Sources/NIOHTTP1/HTTPDecoder.swift]: whilst we're on it, mind also renaming that `errno` variable up there?

**Comment by weissi** [comment]
@Lukasa compilation failed though:

```
19:01:45 Linking ./.build/x86_64-unknown-linux/debug/NIOWebSocketServer
19:01:55 /code/Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift:155:32: error: value of optional type '[UInt8]?' not unwrapped; did you mean to use '!' or '?'?
19:01:55         XCTAssertEqual(handler.body, Array("some body data".utf8))
19:01:55                                ^
19:01:55                                    !
19:01:55 /code/Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift:166:32: error: value of optional type '[UInt8]?' not unwrapped; did you mean to use '!' or '?'?
19:01:55         XCTAssertEqual(handler.body, Array("some body data".utf8))
19:01:55                                ^
19:01:55                                    !
```

**Comment by Lukasa** [comment]
@jw Ok, that should be addressed.
